### ISSUE TITLE
When fit_lc is passed an object that is already a PhotometricData object, with modelcov=True, the object is modified in-place

### ISSUE THREAD

**Issue by drhay53** [sncosmo/sncosmo#215]
I know from my other issue that users are not meant to use the PhotometricData object. I was using one due to the ease of slicing the object when a covariance matrix is present (now I know I should be using select_data for that). 

Anyway, when fit_lc is passed a PhotometricData object, it silently accepts it, and modifies the object in memory while it is running.

With modelcov=True, this means that the data object that is passed to fit_lc is modified silently to include the model covariance; subsequent calls to fit_lc then add more model covariance, etc. I'm not sure if this is the right parlance, but usually I feel "namespace-safe" in passing an object to a function, in that the object will not be modified silently. 

While users should probably not be using the PhotometricData class directly, would it perhaps be safer if the constructor of the object returned a copy, if it is already a PhotometricData object? Along with the other PR, I can make the change at some point, but I'm just not sure the "right" way to do it, so let me know if there is any action item here.

**Comment by drhay53** [comment]
So, I changed back to using tables as the data to pass to fit_lc, and the data table is still being modified unless I pass fit_lc a deepcopy of the data. Unfortunately, however, I cannot reproduce this with a minimal working example. 

I am dumping the covariance matrix of the data to a file before and after the call to fit_lc with modelcov=True. Before fit_lc, the covariance matrix is diagonal only. After fit_lc, it has off-diagonal terms. 

If I pass fit_lc a deepcopy, obviously nothing is modified.

Python strikes again...

**Comment by kbarbary** [comment]
I'm not sure if this is the entire issue, but [this line](https://github.com/sncosmo/sncosmo/blob/master/sncosmo/fitting.py#L37) should be `cov = cov + mcov`, not `cov += mcov`. The way it is written now, `data.fluxcov` is being modified in-place, which was not intended.

Can you try that and see if it fixes all the issues you see?

**Comment by rbiswas4** [comment]
@drhay53  I made the changes @kbarbary suggested and can issue a PR. (I think that is important either way and happens in two places : `generate_chisq` and `chisq`). 

However, I am not able to reproduce your error. Here is what I tried (on a `sncosmo` *without any changes*) : 

```
import sncosmo
import numpy
from copy import deepcopy

data = sncosmo.load_example_data()
cov = np.diag(data['fluxerr']**2)
# This has the same fluxerr for all points
cov[9, 10] = 0.5 # These have SNR > 10, so will not be skipped through minsnr
cov[9, 10] = 0.5
data['fluxcov'] = cov

d == deepcopy(data)
sncosmo.fit_lc(data, model, ['z', 'x0', 'x1', 'c', 't0'], bounds=dict(z=(0.3, 0.7)), modelcov=True) 
d == data
```
This seems to work fine and returns `True`. I am guessing , so I am wondering if you can put your example data and covariance, because it seems there is more to it that I might not be getting.

**Comment by drhay53** [comment]
I also tried to make a minimal working example and was unable to reproduce
it.

I am fitting JLA with a custom class that sets up the data, does phase cuts
with my own procedure for estimating time of maximum, and has a custom
chi^2 procedure that handles the model covariance. (Much of this
infrastructure was setup with older versions of sncosmo and was, at the
time, necessary).

I found this issue by noticing that the SDSS light curves in my sample had
larger errors than with previous sncosmo versions. The SNLS, lowz, and
RiessHST samples seemed to have the same uncertainties in the light curve
fits. In my phase cuts, there are two calls to fit_lc, and the SDSS SNe
were getting the model covariance added to the data in each fit, prior to
me separately adding it in. So the SDSS SNe were getting the model
covariance added 3 times.

It's unclear to me why only the SDSS SNe in my procedure are affected in
the final light curve uncertainties. Other than an initial setup of the
covariance matrix, which looks for the covmat files (only included in JLA
for SNLS), there is no distinction in the code for which sample the light
curve comes from.

The data that is being passed to fit_lc is an attribute of the class that
handles all of the fitting (i.e., self.data is passed to fit_lc). I even
tried setting up an MWE where the data were part of a class, and still
didn't see the behavior.

However, I am completely certain that fit_lc is, in some scenarios,
modifying the data in place, because I have dumped the covariance matrix to
file before and after fit_lc and observed off-diagonal terms appearing. By
passing a deepcopy of the data to fit_lc, my uncertainties on SDSS SNe are
now in agreement with my previous sncosmo 1.2.0 runs. So I am pretty
confident that even if the data are passed to fit_lc as an astropy table,
in some cases, they are modified in-place.

Brian Hayden
Lawrence Berkeley National Laboratory
050-5040, 510-486-4169

**Comment by rbiswas4** [comment]
> It's unclear to me why only the SDSS SNe in my procedure are affected in
the final light curve uncertainties. 

That is interesting. It might be that the correlations are greater in the SDSS sample than in SNLS. It might also be that some of these happen to be near the edges of the usual parameter space for SNIa, where the model covariance effect is larger. Not quite sure why that would affect SDSS only. 


> However, I am completely certain that fit_lc is, in some scenarios,
modifying the data in place, because I have dumped the covariance matrix to
file before and after fit_lc and observed off-diagonal terms appearing. By
passing a deepcopy of the data to fit_lc, my uncertainties on SDSS SNe are
now in agreement with my previous sncosmo 1.2.0 runs. So I am pretty
confident that even if the data are passed to fit_lc as an astropy table,
in some cases, they are modified in-place.

I think the thing @kbarbary pointed out is certain to do this. So, fixing that is important. It also happens in two places as I pointed out (but am not sure if they are both being used). 

I am wondering what it would take to test that this problem is gone? That was my intention when I tried that script, but evidently that is not capturing some effect of the SDSS SN.

**Comment by rbiswas4** [comment]
That works! A working example even if not a minimal example will do. Feel free to use the version at `https://github.com/rbiswas4/sncosmo` where I added this change in, if that helps. I can make a PR if you guys prefer. (Alternatively if you can attach the particular light curve here, I could help test it out) 

I am hoping for a quick new release of `sncosmo` because of some astropy related new feature that will aid in installation. Hence trying to fix these :)

**Comment by kbarbary** [comment]
@rbiswas4 Perhaps your minimal working example doesn't show the issue because of #213? Can you retry it, but call the covariance column `'cov'`?

**Comment by rbiswas4** [comment]
@kbarbary This sounds possible. On the version with the added changes to the alias: 
I can change `fluxcov` to `cov` and both of them give the right answer. 
On the old version, `fluxcov` works (maybe ignored) but with `cov` I get a failure. 

Here are the results:
On 'f54fd9f'  : 
running:
```python test_script.py```
I get the output
```
/home/rbisw/.local/lib/python3.6/site-packages/sncosmo-1.6.dev0-py3.6-linux-x86_64.egg/sncosmo/models.py:737: RuntimeWarning: invalid value encountered in true_divide
  result = v * (f0 / ftot)**2 * scale**2
[ True False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False]```

While on `813b707`

```
python test_script.py```

I get ```
/Users/rbiswas/.local/lib/python2.7/site-packages/sncosmo-1.6.dev0-py2.7-macosx-10.6-x86_64.egg/sncosmo/models.py:737: RuntimeWarning: invalid value encountered in divide
  result = v * (f0 / ftot)**2 * scale**2
[ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True]
```
when the attached script was used. (The script above was copied by hand and had some errors).
The script worked on both if the name `cov` was replaced by `fluxcov`. This is meaningful in the 813b707 commit, because it fixes #213 , but if it is skipping the column in the older commit, that is a bit of a problem, as I got no indication this was happening. Anyway, since this seems to improve the problem, I am opening a PR with the modifications. It would be good to get a quick confirmation from @drhay53 !

[test_script.py.zip](https://github.com/sncosmo/sncosmo/files/1949866/test_script.py.zip)
